### DISCUSSION TITLE
switchmode 15 and rules

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#16586] [Q&A]
I know there are some subjects already about that but for me no clear answer was given

I have a PIR sensor with the presence signal linked to a pin of the ESP32 with tasmota, and the tasmota drives a pwm for leds on another pin and want to drive all the behavior locally on the tasmota with the rules.

since the PIR sensor sens a long pulse, i assimilate it as a switch more than a button.
i want the relay to be independant from the switch so i can program the behavior with rules
I set switchmode 15 , i use this rule : 
rule 1 : 
ON Switch1#state=1 do power1 1 ENDON
ON power1#state=1 DO RuleTimer1 600 ENDON 
ON Rules#Timer=1 DO Power1 0 ENDON

the first rule never triggers (i independently tested the second and third pat of the rule which are OK).
there is also mentions of setoption 114 but i barely see the difference between both.
Is the lack of triggering normal whil i use switchmode 15  ? 

i know there might be some other methods but this discussion is more to ask : doesnt it looks natural while you want to use the rules to use the function that disconnect the button and the switch in the switchmode list ?

**Comment by sfromis** [comment]
The text for SwitchMode 15 specifically says:
> Send **only** MQTT message on switch change. 

That "only" is meant quite literally, you've told it that you want nothing with rules.

More generally, if you have rules to handle all switch states, you need nothing else to decouple. You could also use `SetOption114` to decouple. Beware that different ways of decoupling needs different triggers. Simplest is just having rules and nothing more.

**Reply by ouinouin** [reply]
literally meaning, it doesnt event emit a trigger ? 
the setoption 114 is working for my need, but for me its a strange behaviour that the switchmode15 is not giving any trigger , i dont see any drawback for this mode to emit one trigger.

**Reply by sfromis** [reply]
It is clearly documented that it is **only** MQTT publish, nothing more. And there is no problem using a switchmode which allows triggers, or simply have rules to cover all the state values, meaning that you do not have to worry about decoupling.

**Reply by Toepolev** [reply]
Hello Ouinouin
I understand your problem because I'm struggeling with the same.
Even with SetOption114 my rules where I want to make decisions based on the state of a Switch (that is decoupled) will not work.
In the answer from sfromis I read that he suggests simply having rules to cover all the state values. The only problem is that de state values will not work in rules when they are decoupled and they have to be because they would else trigger the output directly and not via the rule. 
Did you find a solution for your problem?

**Reply by sfromis** [reply]
If you want to use basic Tasmota rules for handling Switch states, you indeed really do **not** have to use `SetOption114`. When your rules cover the possible state values you can get, you need no "decoupling" at all, and you will **not** get the default actions triggered.

Better not try doing anything for "decoupling", as this is to get MQTT messages instead of default actions, when not using rules. Simplest is a "cover-all" rule like `on Switch2#state do ... endon`. One caveat is that this is about "classic rules", if you use Berry `tasmota.add_rule`, this will not supersede the default actions without classic rules (possibly doing nothing there). 

And do **not** use `SwitchMode` 15/16, as these will **only** give you MQTT messages, not the usual state values for rules.

Also remember that rules will give you no error messages if you do not follow the syntax strictly, but instead just not work.

**Reply by Toepolev** [reply]
Maybe I don't understand it well. 
In my project, I want to control my heatingboiler with a Wemos Mini. The image shows my configuration. 
As I understand it (and also notice), the switch number is directly linked to the relay number in Tasmota (coupled). 
In my case, the burner will start when the boiler thermostat closes, and the circulation pump will start when the room thermostat closes. However, this is **not** my desire.

The conditions for starting the burner are: A closed room thermostat (NO) **AND** a closed burner thermostat (NC). 
The conditions for the circulation pump are a closed room thermostat **AND** a minimum boiler temperature of 40°C. 
If I don't use "decoupling", the pump will start when the room thermostat closes. The boiler will always start when the boiler thermostat closes. 
That should not happen. Closing switch 1 should not directly control relay 1, and closing switch 2 should not directly control relay 2.

Pump = Switch1 **AND** Temp DS1820>40°C 
Burner = Switch1 **AND** Switch2

In the rules I would use: 
PUMP: (IF (Switch1#State=1) AND (DS18B20#Temperature>40)) 
BURNER: ((IF Switch1#State=1) AND (Switch2#State=1))
               or (IF Switch1#State=1) AND (DS18B20#Temperature<50))
For me it seems that I have to decouple anyway but it looks like I can not use the mqqt state in my rules.         

( The "AND" does not seem to work directly in Rules and need the IF Then. 
So Rule1 ON Switch1#State=1 AND DS18B20#Temperature>40 DO Power1 1 ENDON does not work. )
![Wemos ketel](https://github.com/user-attachments/assets/57652d7b-6aff-42cc-a5c4-4637a3f80b98)

**Reply by sfromis** [reply]
Strictly speaking, the Switch gpio index is not **directly** linked a Relay gpio, the coupling is by order, not by number, which only "happens" to match in straightforward configs like yours.

You indeed cannot use `AND` in the trigger part of a rule before `do`, only if you have `IF/ENDIF` in the rule body after `do`, but you can **only** use the trigger expressions with `#` in the trigger, not in the rule body. This is due to triggers being **when** something happens, not about current state.

In some cases, you have [rule variables](https://tasmota.github.io/docs/Rules/#rule-variables) for current state instead of an event, including `%switch1%` etc, but excluding sensors like temperature.

And yes, when going beyond the simplest cases like a button/switch input for the purpose of directly controlling a relay output, you'll often have to avoid the default actions. But you do not "need" to use explicit decoupling (SwitchMode 15/16 or SetOption114) for this, as long as you have rule triggers picking up all possible state values for the input type you have.
